// jsonc-simple-parser/dist/tokenize/tokens.js
"use strict";
/* IMPORT */
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
const context_1 = require("./context");
/* HELPERS */
const makeChild = (type) => (values) => {
    const source = values[0];
    context_1.default.offset += source.length;
    return { type, source };
};
const makeParent = (type) => (values) => {
    const children = values.flat();
    return { type, children };
};
/* TOKENS */
const Tokens = {
    EarlyReturn: () => {
        if (context_1.default.offset > context_1.default.offsetMax)
            return [];
    },
    Insufficient: (values) => {
        if (values[0].length)
            Tokens.Invalid(values);
        throw new SyntaxError('Unexpected end of JSONC input');
    },
    Invalid: (values) => {
        throw new SyntaxError(`Unexpected token ${values[0]} in JSONC at position ${context_1.default.offset}`);
    },
    Passthrough: (values) => {
        return values.flat().filter(utils_1.default.isToken);
    },
    Newline: makeChild('Newline'),
    Whitespace: makeChild('Whitespace'),
    CommentLine: makeChild('CommentLine'),
    CommentBlock: makeChild('CommentBlock'),
    Comma: makeChild('Comma'),
    CommaTrailing: makeChild('CommaTrailing'),
    Colon: makeChild('Colon'),
    Null: makeChild('Null'),
    True: makeChild('True'),
    False: makeChild('False'),
    Number: makeChild('Number'),
    String: makeChild('String'),
    ArrayOpen: makeChild('ArrayOpen'),
    ArrayClose: makeChild('ArrayClose'),
    Array: makeParent('Array'),
    ObjectOpen: makeChild('ObjectOpen'),
    ObjectClose: makeChild('ObjectClose'),
    Object: makeParent('Object'),
    Root: makeParent('Root')
};
/* EXPORT */
exports.default = Tokens;
