// configuration/dist/providers/file.js
"use strict";
/* IMPORT */
Object.defineProperty(exports, "__esModule", { value: true });
const file_1 = require("../utils/file");
const memory_1 = require("./memory");
/* FILE */
class ProviderFile extends memory_1.default {
    constructor(options) {
        super(options);
        this.watching = !!options.watch;
        this.writeOptions = options.writeOptions;
        this.writeSyncOptions = options.writeSyncOptions;
        this.swap(options.path, true);
    }
    dispose() {
        this.unwatch();
    }
    swap(path, _initial = false) {
        if (path === this.path)
            return;
        this.dispose();
        this.path = path;
        this.init();
        if (!_initial)
            this.triggerChange();
        if (this.watching)
            this.watch();
    }
    watch() {
        if (!this.path)
            return;
        const path = this.path;
        this.watcher = file_1.default.watch(path, async () => {
            const { dataRaw } = await this.read();
            if (path !== this.path)
                return;
            if (this.isEqual(dataRaw))
                return;
            super.write(dataRaw, true);
        });
    }
    unwatch() {
        if (!this.watcher)
            return;
        this.watcher.close();
        delete this.watcher;
    }
}
/* EXPORT */
exports.default = ProviderFile;
