// configuration/dist/providers/memory.js
"use strict";
/* IMPORT */
Object.defineProperty(exports, "__esModule", { value: true });
const plain_object_clone_1 = require("plain-object-clone");
const path_prop_1 = require("path-prop");
const type_1 = require("../utils/type");
const abstract_1 = require("./abstract");
/* MEMORY */
class ProviderMemory extends abstract_1.default {
    async read() {
        return this.readSync();
    }
    readSync() {
        var _a, _b;
        const data = (_a = this.data) !== null && _a !== void 0 ? _a : plain_object_clone_1.default(this.defaults), dataRaw = (_b = this.dataRaw) !== null && _b !== void 0 ? _b : this.defaultsRaw;
        return { data, dataRaw };
    }
    async write(data, force = false) {
        return this.writeSync(data, force);
    }
    writeSync(data, force = false) {
        var _a, _b;
        if (!force && this.isEqual(data))
            return;
        if (type_1.default.isString(data)) {
            this.data = path_prop_1.default.unflat((_a = this.dataParser.parse(data)) !== null && _a !== void 0 ? _a : this.defaults);
            this.dataRaw = data;
            this.dataSchema = this.filterer(this.data);
        }
        else {
            this.data = path_prop_1.default.unflat(data);
            this.dataRaw = (_b = this.dataParser.stringify(data, this.dataRaw)) !== null && _b !== void 0 ? _b : this.defaultsRaw;
            this.dataSchema = this.filterer(this.data);
        }
        this.triggerChange();
    }
}
/* EXPORT */
exports.default = ProviderMemory;
