// watcher/dist/constants.js
"use strict";
/* IMPORT */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RENAME_TIMEOUT = exports.POLLING_TIMEOUT = exports.POLLING_INTERVAL = exports.PLATFORM = exports.MAX_SAFE_INTEGER = exports.IS_WINDOWS = exports.IS_MAC = exports.IS_LINUX = exports.HAS_NATIVE_RECURSION = exports.DEPTH = exports.DEBOUNCE = void 0;
const os_1 = __importDefault(require("os"));
/* CONSTANTS */
const DEBOUNCE = 300;
exports.DEBOUNCE = DEBOUNCE;
const DEPTH = 20;
exports.DEPTH = DEPTH;
const { MAX_SAFE_INTEGER } = Number;
exports.MAX_SAFE_INTEGER = MAX_SAFE_INTEGER;
const PLATFORM = os_1.default.platform();
exports.PLATFORM = PLATFORM;
const IS_LINUX = (PLATFORM === 'linux');
exports.IS_LINUX = IS_LINUX;
const IS_MAC = (PLATFORM === 'darwin');
exports.IS_MAC = IS_MAC;
const IS_WINDOWS = (PLATFORM === 'win32');
exports.IS_WINDOWS = IS_WINDOWS;
const HAS_NATIVE_RECURSION = IS_MAC || IS_WINDOWS;
exports.HAS_NATIVE_RECURSION = HAS_NATIVE_RECURSION;
const POLLING_INTERVAL = 3000;
exports.POLLING_INTERVAL = POLLING_INTERVAL;
const POLLING_TIMEOUT = 20000;
exports.POLLING_TIMEOUT = POLLING_TIMEOUT;
const RENAME_TIMEOUT = 1250;
exports.RENAME_TIMEOUT = RENAME_TIMEOUT;
