// watcher/dist/utils.js
"use strict";
/* IMPORT */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const are_shallow_equal_1 = __importDefault(require("are-shallow-equal"));
const fs_1 = __importDefault(require("atomically/dist/utils/fs"));
const debounce_1 = __importDefault(require("debounce"));
const path_1 = __importDefault(require("path"));
const tiny_readdir_1 = __importDefault(require("tiny-readdir"));
const constants_1 = require("./constants");
/* UTILS */
const Utils = {
    lang: {
        areShallowEqual: //TODO: Import all these utilities from "nanodash" instead
        are_shallow_equal_1.default,
        debounce: debounce_1.default,
        attempt: (fn) => {
            try {
                return fn();
            }
            catch (error) {
                return Utils.lang.castError(error);
            }
        },
        castArray: (x) => {
            return Utils.lang.isArray(x) ? x : [x];
        },
        castError(exception) {
            if (Utils.lang.isError(exception))
                return exception;
            if (Utils.lang.isString(exception))
                return new Error(exception);
            return new Error('Unknown error');
        },
        defer: (callback) => {
            return setTimeout(callback, 0);
        },
        isArray: (x) => {
            return Array.isArray(x);
        },
        isError(x) {
            return x instanceof Error;
        },
        isFunction: (x) => {
            return typeof x === 'function';
        },
        isNumber: (x) => {
            return typeof x === 'number';
        },
        isString: (x) => {
            return typeof x === 'string';
        },
        isUndefined: (x) => {
            return x === undefined;
        },
        noop: () => {
            return;
        },
        uniq: (arr) => {
            if (arr.length < 2)
                return arr;
            return Array.from(new Set(arr));
        }
    },
    fs: {
        isSubPath: (targetPath, subPath) => {
            return (subPath.startsWith(targetPath) && subPath[targetPath.length] === path_1.default.sep && (subPath.length - targetPath.length) > path_1.default.sep.length);
        },
        poll: (targetPath, timeout = constants_1.POLLING_TIMEOUT) => {
            return fs_1.default.statRetry(timeout)(targetPath, { bigint: true }).catch(Utils.lang.noop); // Under Windows we need stats objects using bigints https://github.com/nodejs/node/issues/12115
        },
        readdir: async (rootPath, ignore, depth = Infinity, signal) => {
            const result = await tiny_readdir_1.default(rootPath, { depth, ignore, signal });
            return [result.directories, result.files];
        }
    }
};
/* EXPORT */
exports.default = Utils;
